skyportal/facility_apis/atlas.py

```diff
@@ -16,7 +16,7 @@
 
 env, cfg = load_env()
 
-if cfg['app.atlas.port'] is None:
+if cfg.get('app.atlas.port') is None:
```

Comment:
Correct fix is to define ATLAS port in the default configuration file, as per https://github.com/skyportal/skyportal/pull/3091

Reply:
In that case we can just skip the entire conditional.

Reply:
Not really, it's fine to have the port be None.

Reply:
If it is fine to have it None, then you must use `cfg.get()`.

Reply:
No, that's not accurate.  `cfg.get` is for when keys are not defined in the configuration file, just like with dics: you use `.get` when they key may be missing.  A key may have a *value* of None without being missing.

E.g.:

```
server:
  port:
```

`cfg["server.port"]` is valid, and is None.

If you wanted to parse, instead:

```
server:
  url: http://...
```

Then you'd have to use `cfg.get("server.port")`, since `port` is undefined.

Reply:
Yeah I get that. I just feel like this sort of subtlety doesn't help us much. For example, a casual sysadmin will see this config file and delete the empty port and then wonder why nothing works. Whereas we can be nice to the poor guy and give him a reasonable default in the code. 

I'm not sure what I think about having defaults defined in the config and in the code. On the one hand defining it in two places leaves us open to inconsistencies (but this already happens since we have the same things defined in multiple config files). On the other hand, this seems like a ground-floor definition in the code that will be there for you if someone decides to delete stuff they don't care about in the config. But it will also be quietly overridden by any config file(s) above that level.

Reply:
I don't understand: where will they delete it?

Reply:
I don't know if that would help. Is the default config even used once you create a `config.yaml`?

Reply:
It is, by default:

```
    # Always load the default configuration values first, and override
    # with values in user configuration files
    all_configs = [
        Path(basedir / "config.yaml.defaults"),
        Path(basedir / "../config.yaml.defaults"),
    ] + config_files
    all_configs = [cfg for cfg in all_configs if os.path.isfile(cfg)]
    all_configs = [os.path.abspath(Path(c).absolute()) for c in all_configs]

    cfg = Config(all_configs)
```

Reply:
From staging:

```
Config files: baselayer/config.yaml.defaults
                config.yaml.defaults
                config.yaml
                /etc/skyportal/config.yaml
```

Reply:
I see, but that still doesn't help if you delete a sub-key as in `atlas.port`.

Reply:
Besides, having defaults, even easy ones like `None`, are better than throwing an exception in most cases.

Reply:
How do you delete subkeys? I don't think that's possible with our current setup.

I think a default of None is not helpful. As we've seen, it hides errors when you think you are accessing keys when they don't exist (I had to fix instances of this in my PR).

Reply:
If you leave the key `atlas` and in the new config file have `atlas` with no `port` sub-key, then `cfg['atlas.port']` will send an error. If indeed there is no way to use `atlas` without a `port` then I agree there should not be a default, and then there should not be a conditional checking if it is `None`. 
On the other hand, if in some cases you don't need a `port` I don't see any harm allowing the user to either leave an empty `port` sub-key or deleting it altogether. In that case the default is `None`, you check for it (as in the way the code worked until now) and decide to use it or not. 

In other cases like spectrum types, it's not a big problem to have a hard coded default. I think that's better than having an error caused by a missing config key that's not critical to what you are working on. I could be not using spectra at all in my application, and I could delete the `spectrum_type` key and then spend hours figuring out what is `spectrum_types` and why it's complaining. Or I could just have reasonable defaults hard coded so I never have errors there. But that's just me...

Reply:
The configuration works by overriding a dictionary, so if you do not specify the atlas port it is whatever we define it to be in our default config.

The only place I think it's possible to remove entries is where you specify a list. If you specify an empty list, obviously entries from the old list would not be visible.

It is *perfectly fine* to make the code robust against cases where the configuration may be entirely missing.  In those cases (see the weather for example), just use `cfg.get` explicitly.  Again, I think it won't be necessary very often, but the option is there.

Reply:
OK. I think we are basically in agreement. Although I think the missing key will be missing in the atlas case because it is a key under another key. 

I also think using `get` is better unless you really must have something configured (then an error is the right behavior). I just feel like that's more common than you think. In any case, that requires manually moving `cfg[..]` to `cfg.get(..)` in all the places, which I don't want to do in this PR (this already touches way too many files).

Reply:
We do recursive updates. Try it out, maybe?